Tensor construction warnings

meteore opened the pull request:
This PR adds a little helper function to prevent tensor construction warnings when passing a `torch.Tensor`-typed argument to `torch.tensor()`.

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -41,14 +41,14 @@ def _log_scale(self):
     def forward(self, inputs, context=None):
         batch_size = inputs.shape[0]
-        num_dims = torch.prod(torch.tensor(inputs.shape[1:]), dtype=torch.float)
+        num_dims = torch.prod(ensure_tensor(inputs.shape[1:]), dtype=torch.float)
```
I think it's safe to assume `inputs` will be a tensor, so no `ensure_tensor` needed. Doubt many other transforms would work with non-tensor inputs, and it's a good convention to keep.

Same in `inverse()`.

meteore replied:
Great, I will revert those, but then also no wrapping in `torch.tensor`?. 

Shall we keep then ensure_tensor for the other (admittedly few) cases?

arturbekasov replied:
Oh, I've misread the code. We're converting the tensor shape to a tensor, not the inputs themselves -- sorry! I think we should avoid tensors all-together here and do `np.prod(inputs.shape[1:])` instead to compute `num_dims`.

arturbekasov replied:
> Shall we keep then ensure_tensor for the other (admittedly few) cases?

Don't mind either way. Less code is good, but if you think it could be useful in other places at some point: feel free to leave it.

meteore replied:
There are only few spots where nflows depends on numpy so here a few suggestions that sidestep the dependency (and perhaps are more clear; I am not sure `num_dims` is the right name for the number of elements in a batch).

If assuming `inputs` is a tensor is safe, we could do [`inputs.numel()`](https://pytorch.org/docs/stable/torch.html#torch.numel) or, in this particular situation where we want the number of elements of all but the first dimension, `inputs[0].numel()` or even `inputs.numel()//batch_size`.  

If you still want to explicitly work on the shapes (vs. using `.numel()`), we could do this:
`batch_size, *batch_shape = inputs shape`
`batch_numel = np.prod(batch_shape)`.

Finally, as an interesting note on what Python version to require Python 3.8 offers `math.prod` (, apart from `from reduce import functools; from operator import mul; reduce(mul, inputs.shape)`, which is always possible).

meteore replied:
Don't want to drown you in options; just trying to get an understanding of stylistic preferences, since of some of these idioms would be applicable elsewhere ;).

arturbekasov replied:
> There are only few spots where nflows depends on numpy

I don't think a dependency on such a widely used tool as `numpy` should worry us at all at this point, especially if it makes for more concise/readable code.

> I am not sure `num_dims` is the right name for the number of elements in a batch).

Agreed. `batch_numel` that you propose is better.

> `inputs[0].numel()` or even `inputs.numel()//batch_size`

I think both are more cryptic than necessary.

> `batch_size, *batch_shape = inputs.shape`
> `batch_numel = np.prod(batch_shape)`.

This works too -- very Pythonic! 🙂

> Finally, as an interesting note on what Python version to require Python 3.8 offers `math.prod`

I can't believe it has taken them this long.

meteore commented:
Re. `AffineTransform` replacing `AffineScalarTransform` the test fails on the `logabsdet` of the inverse in `standard_test.py::AffineScalarTransformTest`. I could have a look tomorrow, if you don't immediately see the reason.

meteore commented:
Ok, the commits above implement all the changes we've discussed so far
* typing
* unifying `AffineScalarTransform` and `AffineTransform` under `PointwiseAffineTransform`.
* more explicit (IMHO) treatment of defaults, but compatibility classes still allow passing None (e.g. the tests use the old interface).
* some more goodies.

Note I had to special-case scalar `scale` to match the expectation of the tests to within `eps=1E-6`. Using the `.sum()` idiom leads to disagreements `~1E-5`, and is perhaps slower. But now the special casing is confined to a little function, where it can be suitably documented. Also, this function could be memoized if the transform is applied multiple times (or even in fwd and inv modes) with the same `batch_shape`.

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -1,82 +1,92 @@
-        batch_size = inputs.shape[0]
-        logabsdet = torch.zeros(batch_size)
+    def forward(self, inputs: torch.Tensor, context=Optional[torch.Tensor]):
+        num_batches = inputs.shape[0]
```
We should probably stick to `batch_size` for consistency. Also,  `num_batches` doesn't make sense to me: we assume `inputs` will be a single batch?

Same in `PointwiseAffineTransform`.

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -1,82 +1,92 @@
-        if scale == 0.0:
-            raise ValueError("Scale cannot be zero.")
+        # reject scales < ~0 (upto dtype precision)
+        is_scale_positive = torch.isfinite(torch.log(scale)).any()
```
Why not a much simpler `(scale <= 0).any()`?

meteore replied:
That's right! This is a mental leftover from `scale` potentially being a scalar. We had then:  
> In [15]: (1E-46>0, tensor(1E-46)>0)                                                                 
Out[15]: (True, tensor(False))

and I wanted to test closest to the point of (potential) breakage - conservatively. But with `scale` a tensor, this no longer makes sense.

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -1,82 +1,92 @@
+class PointwiseAffineTransform(Transform):
+    def __init__(
+        self,
+        shift: torch.Tensor = torch.tensor(0.0),
```
1. Wouldn't the correct type be `Union[torch.Tensor, float]`?
2. I agree that explicit defaults (rather than `None`) are better. But my internal alarm is going off when I see a mutable default value. Can we change to scalar (hence immutable) defaults?

meteore replied:
Thanks for catching this, agree wholeheartedly!

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -1,82 +1,92 @@
+            return self._log_scale.expand(batch_shape).sum()
+        else:
+            # optimise for scalar scale: expand = numel equal scalars
+            # also needed to match test at eps < 1E-5
```
I think the comment could be more informative than "This is to make the test pass": computing `x*y` is more numerically accurate than computing the same thing by adding `x` `y` times: we lose precision on each addition.

meteore replied:
Absolutely, thanks! 

Related, two lines above we're now summing after expanding to `batch_shape` but _before_ expanding to `batch_size`, which should bring us some extra efficiency. For convenience, here is the previous version

>  logabsdet = self._log_scale.reshape(1, -1).repeat(batch_size, 1).sum(dim=-1)

vs. now (two different locations):
> return self._log_scale.expand(batch_shape).sum() # in `_batch_logabsdet`
> logabsdet = self._batch_logabsdet(batch_shape).expand(batch_size) # in `forward`

Some more easy efficiency gains could come from memoizing `_batch_logabsdet` for different `batch_shape`s, but I'll leave that for now.

meteore commented:
Thanks, Artur, and please keep the nitpicks coming, I love them!

In this installment
* I implemented your suggestions (but see comments)
* I switched to using `.size()` instead of `.shape`, since a) we are demanding `inputs` to be tensors and b) in preparation for [named tensors](https://pytorch.org/docs/stable/named_tensor.html), which will be supported by `.size()`, but I reckon not by `.shape`.
* It's a bit sad that `torch` did not follow numpy terminology here, because as a consequence this reads slightly confusing:
> batch_size, *batch_shape = inputs.size()
* what do you think instead of
> batch_len, *batch_size = inputs.size()
* for readability I encapsulated the `prod` dance in a `numel` function in torchutils that works with Tensors or Tensor Sizes.
* Recovered docstring that I had let slip inadvertently. We could add some more text there. In general, what do you think of systematically adopting the notation in [the recent review by George et al.](https://arxiv.org/abs/1912.02762)? That would have us use Z instead of X. We could give that paper in the `README.md` as entry point to understand the package.
* in `torchutils` I reimplemented `notinfnotnan` using `torch.isfinite`. Since the idiom is so short and self-explaining, and there are no uses in the code that I could find, I actually would suggest to remove the function.

meteore commented:
* Dispensing with `ensure_tensor` since `torch.as_tensor` does what we want.
* Ditto for `numel`. Caveat: `<my_tensor>.numel()` and `<my_size>.numel()` do each the right thing, but you have to ensure then that you have a size.
* Added return types for methods
* More unspecific return type of `notinfnotnan`, as `Tensor` (I learnt) cannot be subscripted. We could solve this defining own types, but postponing this since `notinfnotnan` does not seem necessary anymore.

arturbekasov commented on tests/utils/torchutils_test.py:
```diff
@@ -1,6 +1,7 @@
 """Tests for the PyTorch utility functions."""
 
 import unittest
+import numpy as np
```
Unused import?

arturbekasov commented on nflows/utils/torchutils.py:
```diff
@@ -1,5 +1,6 @@
 """Various PyTorch utility functions."""
 
+from typing import Union
```
Unused import?

arturbekasov commented on nflows/utils/torchutils.py:
```diff
@@ -160,8 +161,8 @@ def get_temperature(max_value, bound=1 - 1e-3):
 
-def notinfnotnan(x):
-    return torch.all(~torch.isnan(x)) and torch.all(~torch.isinf(x))
+def notinfnotnan(x: torch.Tensor) -> torch.Tensor:
```
Nice: `isfinite` is a good way to do this. Happy for uses of `notinfnotnan` to be replaced with it at some point.

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -1,82 +1,89 @@
+        if self._log_scale.numel() > 1:
+            return self._log_scale.expand(batch_shape).sum()
+        else:
+            # when log_scale is a scalar, we use n*log_scale, which is more
```
Minor: in the project we've tried to keep comments formatted sentences, i.e. capitalized and ending with a period.

arturbekasov commented on nflows/transforms/standard.py:
```diff
@@ -1,82 +1,89 @@
+        self, shift: Union[Tensor, float] = 0.0, scale: Union[Tensor, float] = 1.0,
+    ):
         super().__init__()
+        shift, scale = map(torch.as_tensor, (shift, scale))
```
Nice: didn't know about `as_tensor`.
